resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/kubernetes/integrationtest/KubernetesTestComponents.scala

```diff
@@ -69,4 +76,79 @@ private[spark] class KubernetesTestComponents {
       .set("spark.testing", "false")
       .set(WAIT_FOR_APP_COMPLETION, false)
   }
+
+  def getService[T: ClassTag](
+    serviceName: String,
+    namespace: String,
+    servicePortName: String,
+    servicePath: String = ""): T = synchronized {
+    val kubernetesMaster = s"${defaultClient.getMasterUrl}"
+    val httpClient = getHttpClient(kubernetesClient.asInstanceOf[BaseClient])
+
+    val url = s"${
+      Array[String](
+        s"${kubernetesClient.getMasterUrl}",
+        "api", "v1", "proxy",
+        "namespaces", namespace,
+        "services", serviceName).mkString("/")
+    }" +
+      s":$servicePortName$servicePath"
+    val userHome = System.getProperty("user.home")
+    val kubernetesConf = kubernetesClient.getConfiguration
+    val sslContext = SSLUtils.sslContext(kubernetesConf)
+    val trustManager = SSLUtils.trustManagers(kubernetesConf)(0).asInstanceOf[X509TrustManager]
+    HttpClientUtil.createClient[T](Set(url), 5, sslContext.getSocketFactory, trustManager)
+  }
+
+  def getHttpClient(client: BaseClient): OkHttpClient = {
+    val field = classOf[BaseClient].getDeclaredField("httpClient")
+    try {
+      field.setAccessible(true)
+      field.get(client).asInstanceOf[OkHttpClient]
+    } finally {
+      field.setAccessible(false)
+    }
+  }
+}
+
+object TestBackend extends Enumeration {
+  val SingleNode, MultiNode = Value
+}
+
+object KubernetesClient {
+  var defaultClient: DefaultKubernetesClient = _
+  var testBackend: TestBackend.Value = _
```

Comment:
I believe that the polymorphic test architecture could belong in the scope of this PR - if we're moving towards multiple back ends, it would be nice if we captured the overall architecture on this pass. If we are concerned that such a change would create too large of a diff, can we create a separate PR that merges into this one that implements the polymorphic architecture, merge said separate PR first into this branch, and then merge this PR?

This is primarily to ease review. Much of what could be commented on for this PR could be contingent on whether or not a polymorphic approach is used.